### PR TITLE
fix issue #17

### PR DESCRIPTION
**Pull request by xavierpriour**
#17 was caused by a randomly occuring issue in checkServer: in *some* circumstances (sorry, did not manage to devise a reproducible test), `http.request({method: 'HEAD',...})` returns a statusCode of 2 but also an error ('HPE_INVALID_CONSTANT').

As a result this.async was called **twice** (once in the original request callback, and then again in the request, initiated by the error handler).
I suppose grunt interpreted the second call as the end of the following task (watch) if it had already starded, and ended the build.

I added a check so this.async is only ever called once. I also did a little renaming and used false/true value to signal task failure to grunt.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,3 @@
 node_modules
+#IDE
+.idea
```
Comment (sindresorhus) [inline .gitignore]: nope

**Comment by sindresorhus** [comment]
You can use https://github.com/sindresorhus/onetime to make sure the function is only called once.

File: tasks/php.js [hunk 2]
```diff
@@ -19,30 +20,41 @@ module.exports = function (grunt) {
...
 			}).end();
 		}, 50);
 	}
 
 	grunt.registerMultiTask('php', 'Start a PHP-server', function () {
-		var cb = this.async();
+		checkServerTries = 0;
+		serverStarted = false;
+
+		var done = this.async();
```
Comment (kevva) [inline tasks/php.js]: Why rename?

File: tasks/php.js [hunk 3]
```diff
@@ -58,7 +70,7 @@ module.exports = function (grunt) {
 		getPort(function (err, port) {
 			if (err) {
 				grunt.warn(err);
-				cb();
+				done(false);
```
Comment (kevva) [inline tasks/php.js]: ?

File: tasks/php.js [hunk 4]
```diff
@@ -8,8 +8,9 @@ var objectAssign = require('object-assign');
 
 module.exports = function (grunt) {
 	var checkServerTries = 0;
+	var serverStarted = false;
 
-	function checkServer(hostname, port, path, cb) {
+	function checkServer(hostname, port, path, done) {
```
Comment (kevva) [inline tasks/php.js]: Why rename?

File: tasks/php.js [hunk 5]
```diff
@@ -19,30 +20,41 @@ module.exports = function (grunt) {
 			}, function (res) {
 				var statusCodeType = Number(res.statusCode.toString()[0]);
 				if ([2, 3, 4].indexOf(statusCodeType) !== -1) {
-					return cb();
+					if (!serverStarted) {
+						serverStarted = true;
+						done(true);
+					}
+					return;
```
Comment (kevva) [inline tasks/php.js]:
```js
}

return
```

File: tasks/php.js [hunk 6]
```diff
@@ -19,30 +20,41 @@ module.exports = function (grunt) {
...
 				} else if (statusCodeType === 5) {
 					grunt.fail.warn(
 						'Server docroot returned 500-level response. Please check ' +
 						'your configuration for possible errors.'
 					);
-					return cb();
+					if (!serverStarted) {
+						serverStarted = true;
+						done(true);
+					}
```
Comment (kevva) [inline tasks/php.js]:
```js
}

return
```

File: tasks/php.js [hunk 7]
```diff
@@ -19,30 +20,41 @@ module.exports = function (grunt) {
...
+				checkServer(hostname, port, path, done);
 			}).on('error', function (err) {
 				// back off after 1s
-				if (++checkServerTries > 20) {
-					return cb();
+				if (!serverStarted) {
+					if (++checkServerTries > 20) {
+						return done(false);
+					}
+					grunt.verbose.writeln('PHP server not started. Retrying...');
```
Comment (kevva) [inline tasks/php.js]:
```js
}

grunt.verbose.writeln('PHP server not started. Retrying...');
```

File: tasks/php.js (hunk 4 above)
Comment (xavierpriour) [inline tasks/php.js]: To make it clearer what this does (= signal to grunt that the task is done), I had trouble understanding waht `cb` meant.
I used the same name as grunt docs: http://gruntjs.com/api/inside-tasks

File: tasks/php.js (hunk 6 above)
Comment (xavierpriour) [inline tasks/php.js]: I don't understand - why a new line and no ;  ?

There was no blank line and a ; when it was `return cb();`, and there is none in binVersionCheck either.

File: .gitignore (hunk 1 above)
Comment (xavierpriour) [inline .gitignore]: ok, I'll remove it

File: tasks/php.js (hunk 6 above)
Comment (arthurvr) [inline tasks/php.js]: > I don't understand - why a new line and no ; ?

Consistency and readability of the code. We do want a semicolon though.

Comment (xavierpriour) [inline tasks/php.js]: OK - added a jscs rule to have a newline after block, is that the correct thing?
http://jscs.info/rules.html#requirepaddingnewlinesafterblocks

**Comment by xavierpriour** [comment]
@sindresorhus regarding using `onetime`, I find that having a variable and checking against it:

- allows us to skip a second request in the error handler
- makes the intent clear

On the other hand, calling it twice will blow up the grunt chain and we have no test to protect against it.
So I used onetime in its version that throws an exception.

File: tasks/php.js (hunk 4 above)
Comment (kevva) [inline tasks/php.js]: cb = callback

File: tasks/php.js (hunk 6 above)
Comment (kevva) [inline tasks/php.js]: Not in this PR, no.

Comment (xavierpriour) [inline tasks/php.js]: I'm unclear - is format ok now with the new commit?

Comment (kevva) [inline tasks/php.js]: The JSCS rule should be added in another PR.

Comment (xavierpriour) [inline tasks/php.js]: why?

On Sat, Apr 11, 2015 at 11:50 PM, Kevin Mårtensson <notifications@github.com
> wrote:

> In tasks/php.js
> <https://github.com/sindresorhus/grunt-php/pull/63#discussion_r28198728>:
>
> >  				} else if (statusCodeType === 5) {
> >  					grunt.fail.warn(
> >  						'Server docroot returned 500-level response. Please check ' +
> >  						'your configuration for possible errors.'
> >  					);
> > -					return cb();
> > +					if (!serverStarted) {
> > +						serverStarted = true;
> > +						done(true);
> > +					}
>
> The JSCS rule should be added in another PR.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sindresorhus/grunt-php/pull/63/files#r28198728>.
>



-- 
Xavier Priour

Comment (kevva) [inline tasks/php.js]: Because this PR is about fixing #17.

Comment (arthurvr) [inline tasks/php.js]: Because it's a different topic and unrelated to these changes.

Comment (xavierpriour) [inline tasks/php.js]: Guys, I get it.
But I just fixed a bug, then you and I wasted some time on formatting back-and-forth, so I automated it (one line in .jscsrc). That way next time, everyone saves time.
Same spirit as submitting my PR: win for everyone, bonus continual improvement.
Now if you don't want it, don't take it.

Comment (sindresorhus) [inline tasks/php.js]: @xavierpriour I'm happy to see the rules improved, but PRs/commits should be atomic and focused. The change and code style changes should be a separate PR.

**Comment by sindresorhus** [comment]
@xavierpriour alright

**Comment by sindresorhus** [comment]
Please use a proper and descriptive PR and commit title.

**Comment by xavierpriour** [comment]
Hi @sindresorhus, I understand you have strong requirements on how you want your project code, format, commits and PR to be, and I respect that.
Unfortunately on my side it ends up being very time consuming and frustrating to try and match that, and I can't justify spending more time on that PR.

So I believe the best thing would be for you to take what you want from my code and commit it exactly how you want, then close my PR.

**Comment by sindresorhus** [comment]
It's common convention to follow the existing code style and use proper titles. I thank you for the contribution, but this has taken too much time with too low of a quality to spend more time on.
